reference/reference_test.go

```diff
@@ -0,0 +1,481 @@
+			hostname:   "b.gcr.io",
+			repository: "b.gcr.io/test.example.com/my-app",
+			tag:        "test.example.com",
+		},
+		{
+			input:      "xn--n3h.com/myimage:xn--n3h.com", // ☃.com in punycode
+			hostname:   "xn--n3h.com",
+			repository: "xn--n3h.com/myimage",
+			tag:        "xn--n3h.com",
+		},
+		{
+			input:      "xn--7o8h.com/myimage:xn--7o8h.com@sha512:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff", // 🐳.com in punycode
+			hostname:   "xn--7o8h.com",
+			repository: "xn--7o8h.com/myimage",
+			tag:        "xn--7o8h.com",
+			digest:     "sha512:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
+		},
+		{
+			input:      "foo_bar.com:8080",
+			repository: "foo_bar.com",
+			tag:        "8080",
+		},
+		{
+			input:      "foo/foo_bar.com:8080",
+			hostname:   "foo",
+			repository: "foo/foo_bar.com",
+			tag:        "8080",
+		},
+	}
+	for _, testcase := range referenceTestcases {
+		failf := func(format string, v ...interface{}) {
+			t.Logf(strconv.Quote(testcase.input)+": "+format, v...)
+			t.Fail()
+		}
+
+		repo, err := Parse(testcase.input)
+		if testcase.err != nil {
+			if err == nil {
+				failf("missing expected error: %v", testcase.err)
+			} else if testcase.err != err {
+				failf("mismatched error: got %v, expected %v", err, testcase.err)
+			}
+			continue
+		} else if err != nil {
+			failf("unexpected parse error: %v", err)
+			continue
+		}
+		if repo.String() != testcase.input {
+			failf("mismatched repo: got %q, expected %q", repo.String(), testcase.input)
+		}
+
+		if named, ok := repo.(Named); ok {
+			if named.Name() != testcase.repository {
+				failf("unexpected repository: got %q, expected %q", named.Name(), testcase.repository)
+			}
+			hostname, _ := SplitHostname(named)
+			if hostname != testcase.hostname {
+				failf("unexpected hostname: got %q, expected %q", hostname, testcase.hostname)
+			}
+		} else if testcase.repository != "" || testcase.hostname != "" {
+			failf("expected named type, got %T", repo)
+		}
+
+		tagged, ok := repo.(Tagged)
+		if testcase.tag != "" {
+			if ok {
+				if tagged.Tag() != testcase.tag {
+					failf("unexpected tag: got %q, expected %q", tagged.Tag(), testcase.tag)
+				}
+			} else {
+				failf("expected tagged type, got %T", repo)
+			}
+		} else if ok {
+			failf("unexpected tagged type")
+		}
+
+		digested, ok := repo.(Digested)
+		if testcase.digest != "" {
+			if ok {
+				if digested.Digest().String() != testcase.digest {
+					failf("unexpected digest: got %q, expected %q", digested.Digest().String(), testcase.digest)
+				}
+			} else {
+				failf("expected digested type, got %T", repo)
+			}
+		} else if ok {
+			failf("unexpected digested type")
+		}
+
+	}
+}
+
+func TestSplitHostname(t *testing.T) {
+	testcases := []struct {
+		input    string
+		hostname string
+		name     string
+	}{
+		{
+			input:    "test.com/foo",
+			hostname: "test.com",
+			name:     "foo",
+		},
+		{
+			input:    "test_com/foo",
```

Comment:
Hey guys, thanks for this lib! It's a big improvement for external projects that need to parse Docker references.

I tried to switch [Empire](https://github.com/remind101/empire) over to use distribution/reference instead of our custom types, but run into a problem with this method not returning expected results:

Given this:

```go
n, _ := reference.ParseNamed("remind101/acme-inc")
hostname, _ := reference.SplitHostname(n)
```

I would expect `hostname` to be an empty string, however, the actual value of hostname is `remind101`.

To put it another way, the following test case will fail:

```go
{
        input:    "remind101/acme-inc",
        hostname: "",
        name:     "remind101/acme-inc",
},
```

Is this expected behavior?